uswds/uswds
# USWDS - Bug: Fix invalid aria role in validator list items

## Summary
**Updated the validation component so that users of assistive technologies can receive status updates about the completion of each validation requirement.** This was done by adding a script that generates appropriate ARIA attributes. 

### Details
**Added a script that generates an aria-label for each checklist item.** This allows assistive technologies users to navigate to each checklist item and receive the current status of each validation requirement. 

**Added a script that generates a new span to house the summary of the validation status updates.** This span has `aria-live` and `aria-atomic` attributes so that users of assistive technologies will be notified of status changes. 

**Removed the `aria-checked` attribute from list items in the validation component.** This attribute  is no longer needed now that a text element informs the user if the requirement is complete or incomplete.

**Added `data-validation-complete` and `data-validation-incomplete` attributes so that users can customize copy for `status complete` and `status incomplete` messages.**


## Related issue
Closes: #4905 

Related PRs: This PR incorporates the changes originally merged in https://github.com/uswds/uswds/pull/4719 that were later reverted in https://github.com/uswds/uswds/pull/4906. This PR includes the original update from #4719 and resolves the errors that caused the reversion.

## Preview link
Preview link: [Validation component](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-validation-aria/?path=/story/components-validation--validation)

## Problem statement
### Including the aria-checked attribute generates errors
Adding the `aria-checked` attribute to a `<li>` item results in the following a11y error:
```
Errors in http://localhost:4000/components/validation/:

 • Elements must only use allowed ARIA attributes
   (https://dequeuniversity.com/rules/axe/3.5/aria-allowed-attr?application=axeAPI)

   (#validate-code > li:nth-child(1))

   <li class="usa-checklist__item" data-validator="uppercase"
   aria-checked="false"> Use at least one u...</li>

 • Elements must only use allowed ARIA attributes
   (https://dequeuniversity.com/rules/axe/3.5/aria-allowed-attr?application=axeAPI)

   (#validate-code > li:nth-child(2))

   <li class="usa-checklist__item" data-validator="numerical"
   aria-checked="false"> Use at least one n...</li>
```
These errors reveal that items with the `aria-checked` attribute also need an appropriate role. Reference: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-checked#associated_roles

However, adding `role="checkbox"` to the `<li>` item also returns an error:
```
Detected the following accessibility violations!
       
       1. aria-allowed-role (ARIA role should be appropriate for the element)
       
          For more info, visit https://dequeuniversity.com/rules/axe/4.3/aria-allowed-role?application=axeAPI.
       
          Check these nodes:
       
          - html: <li class="usa-checklist__item" data-validator="uppercase" role="checkbox" aria-checked="true">Use at least one uppercase character</li>
            summary: Fix any of the following:
                       ARIA role checkbox is not allowed for given element
```

### VoiceOver produces a confusing readout
Additionally, VoiceOver generates undesirable and inconsistent readouts. Items can be read twice or only partially. 

## Solution
### aria-checked
The `data-checklist-label` element now informs the keyboard user if the validation requirements are complete or incomplete. There is no longer a need to use an attribute like `aria-checked` to communicate `complete` or `incomplete` to the user. 

### Confusing readout
Changing the following items  generated a more consistent readout in VoiceOver:
- Generated an `aria-label` for each checklist item that includes the original text content plus the item's current status
- Generated a new sr-only span with `aria-live` and `aria-atomic` attributes that houses the summary of all checklist items plus their individual statuses. 

### Tasks completed in this PR:
1. Integrated changes created in `cm-validation-a11y` branch [0684b17](https://github.com/uswds/uswds/pull/4914/commits/0684b17e86b0bd8653edf8874ce94846b7d06e6b)
1. Removed the `aria-checked` attribute from the list item [f077275](https://github.com/uswds/uswds/pull/4914/commits/f0772753cfdce51c64355b745a979635070a162b)
1. Added `tabindex="0"` to make the items in the validation checklist items easier to navigate via tab [29c2b61](https://github.com/uswds/uswds/pull/4914/commits/29c2b618e75eb6e8ec4727323835e93e1f9077dd)
1. Created a new sr-only span element with `aria-live` and `aria-atomic` attributes that houses a status summary [07ad7de](https://github.com/uswds/uswds/pull/4914/commits/07ad7de7dfcd33d179958252706565b07c51284c)
1. Added optional `data-validation-complete` and `data-validation-incomplete` attributes [eeb2e18](https://github.com/uswds/uswds/pull/4914/commits/eeb2e188c324b59fcb0e116b899d114ef15515d8)

## Testing and review
Using the screen readers you have available,
1. Confirm that validation status is discoverable via keyboard navigation . 
1. Confirm that the validation status is read out in a way that makes sense.
1. Confirm that the validation component passes a11y tests. 
1. Add the following data attributes with custom values to `.usa-input` and confirm that the status message in `data-checklist-label` updates accordingly:
    - data-validation-complete
    - data-validation-incomplete
    
Tested on: 
- VoiceOver on Firefox, Safari, Edge for MacOs
- VoiceOver Gestures on iOS

---

Before opening this PR, make sure you’ve done whichever of these applies to you:
